primary/src/block_synchronizer/handler.rs

```diff
@@ -247,4 +258,46 @@ impl<PublicKey: VerifyingKey> Handler<PublicKey> for BlockSynchronizerHandler<Pu
 
         results
     }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn synchronize_block_payloads(
+        &self,
+        certificates: Vec<Certificate<PublicKey>>,
+    ) -> Vec<Result<Certificate<PublicKey>, Error>> {
+        let (tx, mut rx) = channel(certificates.len());
+
+        self.tx_block_synchronizer
+            .send(Command::SynchronizeBlockPayload {
+                certificates,
+                respond_to: tx,
+            })
+            .await
+            .expect("Couldn't send message to block synchronizer");
+
+        // now wait to retrieve all the results
+        let mut results = Vec::new();
+
+        // We want to block and wait until we get all the results back.
+        loop {
+            match rx.recv().await {
+                None => {
+                    debug!("Channel closed when getting results, no more messages to get");
```

Comment:
1. could this be interpreted by someone reading logs as an error case, even when it's the normal termination case for this loop?
2. is the `debug` scope or `trace` more appropriate? (come to think of it, this probably applies to `get_block_headers` above)
3. we're probably going to want to run this task under a timeout. I know we have issues open for this, is there one to which we could tack on taking a look at payload synchronization?

Reply:
I've added the `debug` statement purely for the testing/debug purposes - I wouldn't expect someone to try and use this log in prod (or at least I wouldn't like someone have to worry about something going wrong here on production - should make this robust so they won't have to worry about). That being said, to increase robustness probably having a timeout would make sense here. The downstream component - `block_synchronizer` - is already covered from timeout semantics, but it's true that we can't rely on that only.

Reply:
https://github.com/MystenLabs/narwhal/issues/247